<!DOCTYPE html>
<html lang="zh-hant">
<head>
  <link href='//fonts.googleapis.com/css?family=Lato:300,700,300italic' rel='stylesheet' type='text/css'>
  <link href=//fonts.googleapis.com/css?family=Crimson+Text:400,400italic,600,600italic' rel='stylesheet' type='text/css'>
  <link rel="stylesheet" type="text/css" href="//blog.liang2.tw/theme/css/style.min.css">
  <link rel="stylesheet" type="text/css" href="//blog.liang2.tw/theme/css/pygments.min.css">
  <link rel="stylesheet" type="text/css" href="//blog.liang2.tw/theme/css/font-awesome.min.css">
  <!-- Justfont webfont -->
  <!-- Local development -->
  <!-- <script src="//s3-ap-northeast-1.amazonaws.com/justfont-user-script/jf-35439.js"></script> -->
  <!-- Proudction version v4.9.4 -->
  <script src="//blog.liang2.tw/theme/js/justfont_v4.9.4.js" type="text/javascript" charset="utf-8"></script>
  <link href="//blog.liang2.tw/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="Liang2's blog Atom">

  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <meta name="robots" content="" />
<meta name="author" content="Liang2" />
<meta name="description" content="Notes of fundamental tools and learning resources for handling genomic data in R with Bioconductor." />
<meta name="keywords" content="en, r, bioconductor">
<meta property="og:site_name" content="Liang2's blog"/>
<meta property="og:title" content="Overview of Genomic Data Processing in Bioconductor"/>
<meta property="og:description" content="Notes of fundamental tools and learning resources for handling genomic data in R with Bioconductor."/>
<meta property="og:locale" content="zh_TW"/>
<meta property="og:url" content="//blog.liang2.tw/posts/2015/12/biocondutor-genomic-data/"/>
<meta property="og:type" content="article"/>
<meta property="article:published_time" content="2015-12-29 20:28:00+08:00"/>
<meta property="article:modified_time" content="2015-12-29 20:28:00+08:00"/>
<meta property="article:author" content="//blog.liang2.tw/author/liang2.html">
<meta property="article:section" content="Bioinfo"/>
<meta property="article:tag" content="en"/>
<meta property="article:tag" content="r"/>
<meta property="article:tag" content="bioconductor"/>
<meta name="twitter:card" content="summary">
<meta name="twitter:site" content="@ccwang002">
<meta name="twitter:title" content="Overview of Genomic Data Processing in Bioconductor">
<meta name="twitter:description" content="Notes of fundamental tools and learning resources for handling genomic data in R with Bioconductor.">
  <title>Overview of Genomic Data Processing in Bioconductor &ndash; Liang2's blog</title>
</head>
<body>
  <aside>
    <div>
      <a href="//blog.liang2.tw">
        <img src="//blog.liang2.tw/pics/headpic.jpg" alt="Liang2's Blog" title="Liang2's Blog">
      </a>
      <h1><a href="//blog.liang2.tw">Liang2's Blog</a></h1>
      <p>Code / Stat / Bioinfo</p>
      <nav>
        <ul class="list">
          <li><a href="//blog.liang2.tw/about-me/#about-me">About&nbsp;me</a></li>
          <li><a href="//blog.liang2.tw/talks/#talks">Talks</a></li>
          <li><a href="//blog.liang2.tw/research/#research">Research</a></li>
          <li><a href="//blog.liang2.tw/projects/#projects">Projects</a></li>
        </ul>
      </nav>
      <ul class="social">
        <li><a class="sc-twitter" href="https://twitter.com/ccwang002" target="_blank"><i class="fa fa-twitter"></i></a></li>
        <li><a class="sc-facebook" href="https://www.facebook.com/lbwang.2" target="_blank"><i class="fa fa-facebook"></i></a></li>
        <li><a class="sc-github" href="https://github.com/ccwang002" target="_blank"><i class="fa fa-github"></i></a></li>
        <li><a class="sc-bitbucket" href="https://bitbucket.org/ccwang002" target="_blank"><i class="fa fa-bitbucket"></i></a></li>
        <li><a class="sc-envelope-o" href="mailto:me+blog@liang2.tw" target="_blank"><i class="fa fa-envelope-o"></i></a></li>
        <li><a class="sc-linkedin" href="http://tw.linkedin.com/in/liangbowang/" target="_blank"><i class="fa fa-linkedin"></i></a></li>
      </ul>
    </div>
  </aside>
  <main>
    <nav>
      <a href="//blog.liang2.tw">Home</a>
      <a href="/archives.html">Archives</a>
      <a href="/categories.html">Categories</a>
      <a href="/tags.html">Tags</a>
      <a href="//blog.liang2.tw/feeds/all.atom.xml">Atom</a>
    </nav>

<article>
  <header>
    <h1 id="biocondutor-genomic-data">Overview of Genomic Data Processing in&nbsp;Bioconductor</h1>
    <p>Posted on Dec 29, 2015
      in <a href="//blog.liang2.tw/category/bioinfo.html">Bioinfo</a>
    </p>
  </header>
  <div>
    <p>Sorry for the late update. In the past two months, I finished my Ph.D. applications (hope to hear good news in the next two months) and was busy preparing the PyCon Taiwan 2016. Also, a year-long website development finally came to the&nbsp;end.</p>
<p>Now most things are set so I can back to writing my&nbsp;blog.</p>
<p>Since September, there accumulates at least 5 drafts and I don&rsquo;t know when I can finish them, so I think I have to change my writing strategy. I will first publish things as soon as information collection is done, and deeper reviews will be given in the following posts. Right now I will focus on Bioconductor (and general Bioinformatics topics) and&nbsp;Django.</p>
<div class="toc">
<ul>
<li><a href="#bioconductor">Bioconductor</a><ul>
<li><a href="#annotation-and-genome-reference">Annotation and Genome&nbsp;Reference</a></li>
<li><a href="#experiment-data-storage">Experiment Data&nbsp;Storage</a></li>
<li><a href="#operations-on-genome">Operations on&nbsp;Genome</a></li>
<li><a href="#genomic-data-visualization">Genomic data&nbsp;visualization</a></li>
</ul>
</li>
<li><a href="#summary">Summary</a></li>
</ul>
</div>
<h2 id="bioconductor">Bioconductor</h2>
<p><a href="https://www.bioconductor.org/">Bioconductor</a> is indeed a rich resources for R both in terms of data and tools. And I found I have yet spent time seriously understanding the whole ecosystem, which I believe can drastically lighten the loading of daily&nbsp;analysis.</p>
<p>Bioconductor&rsquo;s website is informative. If you are familar with R, you should already know that in order to understand the usage of a package, one of the best way is to read its vignettes. Packages on Bioconductor generally have vignettes, which is really helpful and the website makes them accessible. On top of that, they have <a href="https://www.bioconductor.org/help/course-materials/">Courses <span class="amp">&amp;</span> Conferences</a> and <a href="https://www.bioconductor.org/help/workflows/">Workflows</a>. The former section collects all conference materials in the past few years, which contains package hands-on, analysis tutorial, and R advanced topics. It&rsquo;s a hidden gem to me since I have already found numerous materials worth reading only after a glance over it. The latter one should be well-known. It gives examples of typical analysis&nbsp;workflows.</p>
<p>I&rsquo;m interested in the following topics in&nbsp;Biocondutor:</p>
<ul>
<li>Annotation and genome reference (OrgDb, TxDb, OrganismDb,&nbsp;BSgenome)</li>
<li>Experiment data storage&nbsp;(ExpressionSets)</li>
<li>Operations on genome&nbsp;(GenomicRanges)</li>
<li>Genomic data visualization (Gviz,&nbsp;ggbio)</li>
</ul>
<p>Keywords in Biocondutors for each topic are attached in the parens, mostly being the package name. For each topic, I&rsquo;ll put the related resources I collected in the following&nbsp;sections.</p>
<p>Before the listing, I found <a href="http://genomicsclass.github.io/book/">PH525x series</a> maintained by Rafael Irizarry and Michael Love from Harvard serves as a comprehensive entry point for almost every related topic. The site is the accompanied resources for their edX classes. Both of them worth taking a&nbsp;look.</p>
<h3 id="annotation-and-genome-reference">Annotation and Genome&nbsp;Reference</h3>
<ul>
<li>
<p><a href="http://genomicsclass.github.io/book/pages/annoPhen.html">Annotating phenotypes and molecular function</a> from <a href="http://genomicsclass.github.io/book/">PH525x series</a> gives a good overview and a taste of the powerful ecosystem Bioconductor&nbsp;provides.</p>
</li>
<li>
<p><a href="https://www.bioconductor.org/help/course-materials/2015/BioC2015/Annotation_Resources.html">Annotation Resources</a> from <a href="https://www.bioconductor.org/help/course-materials/2015/BioC2015/">BioC 2015</a> gives more extensive introduction about all available types of references from genome sequences to transcriptome and gene&nbsp;info.</p>
</li>
</ul>
<p>For example, human comes&nbsp;with</p>
<ul>
<li><a href="http://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html">org.Hs.eg.db</a></li>
<li><a href="http://bioconductor.org/packages/release/data/annotation/html/TxDb.Hsapiens.UCSC.hg38.knownGene.html">TxDb.Hsapiens.<span class="caps">UCSC</span>.hg38.knownGene</a></li>
<li><a href="http://bioconductor.org/packages/release/data/annotation/html/Homo.sapiens.html">Homo.sapiens</a></li>
<li><a href="http://bioconductor.org/packages/release/data/annotation/html/BSgenome.Hsapiens.UCSC.hg38.html">BSgenome.Hsapiens.<span class="caps">UCSC</span>.hg38</a></li>
</ul>
<h3 id="experiment-data-storage">Experiment Data&nbsp;Storage</h3>
<p>ExpressionSet helps store the expression experiment data, which one can combine   expression values and phenotypes of the same sample. Additionally the experiment data (like descriptions of <span class="caps">GEO</span> dataset) can be attached as&nbsp;well.</p>
<ul>
<li>
<p><a href="http://genomicsclass.github.io/book/pages/eset.html">The ExpressionSet container</a> from <a href="http://genomicsclass.github.io/book/">PH525x series</a> gives an intro. It should be sufficient enough to use ExpressionSet in daily&nbsp;work.</p>
</li>
<li>
<p><a href="https://www.bioconductor.org/packages/release/bioc/vignettes/Biobase/inst/doc/ExpressionSetIntroduction.pdf">The ExpressionSet Introduction</a> from its package <a href="https://www.bioconductor.org/packages/release/bioc/html/Biobase.html">Biobase</a>’s vignette gives detailed&nbsp;explanation.</p>
</li>
</ul>
<h3 id="operations-on-genome">Operations on&nbsp;Genome</h3>
<p>I haven&rsquo;t gone into the details, but operations about genomic ranges are often tricky and more importantly, badly&nbsp;optimized.</p>
<ul>
<li>
<p><a href="http://genomicsclass.github.io/book/pages/iranges_granges.html">IRanges and GRanges</a> and <a href="http://genomicsclass.github.io/book/pages/operateGRanges.html">GRanges operations</a> from <a href="http://genomicsclass.github.io/book/">PH525x series</a> give the overview of using the package <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a>.</p>
</li>
<li>
<p><a href="https://www.bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.pdf">An Introduction to Genomic Ranges Classes</a>, a <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a> vignette, gives a detailed&nbsp;view.</p>
</li>
<li>
<p>Also, their paper, <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003118">&ldquo;Software for Computing and Annotating Genomic Ranges&rdquo;, <em><span class="caps">PLOS</span> One</em></a> should be another overview source of the&nbsp;package.</p>
</li>
<li>
<p><a href="https://cran.r-project.org/web/packages/data.table/index.html">data.table</a>&lsquo;s <code>foverlap</code> function worth the comparison, since I already use it and I know it is <a href="https://github.com/Rdatatable/data.table/wiki/talks/EARL2014_OverlapRangeJoin_Arun.pdf">blazingly fast</a>. <code>foverlap</code> handles the overlapping of integer ranges so it can be applied to genomic operation. Its code is quite complex so its mechanism is still a myth to me. I&rsquo;d like to see its comparison with using database like&nbsp;SQLite.</p>
</li>
</ul>
<h3 id="genomic-data-visualization">Genomic data&nbsp;visualization</h3>
<p>Basically I can find two&nbsp;packages:</p>
<ul>
<li><a href="https://bioconductor.org/packages/release/bioc/html/Gviz.html">Gviz</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/ggbio.html">ggbio</a></li>
</ul>
<p>Don&rsquo;t know their difference yet. Both of them can produce well-done figures. But I think I have some experience with ggbio, which was a bit tricky to use. So for now I will go for&nbsp;Gviz.</p>
<ul>
<li><a href="https://www.bioconductor.org/help/course-materials/2012/BiocEurope2012/GvizEuropeanBioc2012.pdf">Visualizing genomic features with the Gviz package</a> given at Bioc Europe 2012 has a decent introduction about&nbsp;Gviz.</li>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/Gviz/inst/doc/Gviz.pdf">The Gviz User Guide</a> looks very comprehensive, which also cover usage with expression and alignment&nbsp;results.</li>
</ul>
<h2 id="summary">Summary</h2>
<p>These resources should be enough for weeks of trying. It&rsquo;s excited to find so many useful&nbsp;tools.</p>
<p>So, good luck to me for my Ph.D. application, PyCon Taiwan 2016, and a shorter blog posting&nbsp;frequency.</p>
  </div>
  <div class="tag-cloud">
    <p>
      <a href="//blog.liang2.tw/tag/en.html">en</a>
      <a href="//blog.liang2.tw/tag/r.html">r</a>
      <a href="//blog.liang2.tw/tag/bioconductor.html">bioconductor</a>
    </p>
  </div>
<div id="disqus_thread"></div>
<script type="text/javascript">
    var disqus_shortname = 'liang2';
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript" rel="nofollow">comments powered by Disqus.</a></noscript>
</article>

    <footer>
<p>
  &copy; Liang2 2015 - This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>
</p>
<p>Built using <a href="http://getpelican.com" target="_blank">Pelican</a> - based on <a href="https://github.com/alexandrevicenzi/flex" target="_blank">Flex</a> theme by <a href="http://alexandrevicenzi.com" target="_blank">Alexandre Vicenzi</a></p><p>
  <a rel="license"
     href="http://creativecommons.org/licenses/by/4.0/"
     target="_blank">
    <img alt="Creative Commons License"
         title="Creative Commons License"
         style="border-width:0"
         src="https://i.creativecommons.org/l/by/4.0/80x15.png"
         width="80"
         height="15"/>
  </a>
</p>    </footer>
  </main>
<script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "BlogPosting",
  "name": "Overview of Genomic Data Processing in&nbsp;Bioconductor",
  "headline": "Overview of Genomic Data Processing in&nbsp;Bioconductor",
  "datePublished": "2015-12-29 20:28:00+08:00",
  "dateModified": "",
  "author": {
    "@type": "Person",
    "name": "Liang2",
    "url": "//blog.liang2.tw/author/liang2.html"
  },
  "image": "//blog.liang2.tw/pics/headpic.jpg",
  "url": "//blog.liang2.tw/posts/2015/12/biocondutor-genomic-data/",
  "description": "Notes of fundamental tools and learning resources for handling genomic data in R with Bioconductor."
}
</script></body>
</html>